doc/monads/monads.lean

```diff
@@ -0,0 +1,307 @@
+operations in a monad can affect other computations in the chain implicitly through side effects,
+especially modification of an implicitly shared value.
+
+## How are monads represented in Lean?
+
+Like functors and applicatives, monads are represented with a type class in Lean:
+
+```lean,ignore
+class Monad (m : Type u → Type v) extends Applicative m, Bind m where
+```
+
+Just as every applicative is a functor, every monad is also an applicative and
+there's one more new base type class used here that you need to understand, namely, `Bind`.
+
+```lean,ignore
+class Bind (f : Type u → Type v) where
+  bind : {α β : Type u} → f α → (α → f β) → f β
+```
+
+The `bind` operator also has infix notation `>>=` where `x >>= g` represents the result of
+executing `x` to get a value of type `f α` then unwrapping the value `α` from that and
+passing it to function `g` of type `α → f β` returning the result of type `f β` where
+`f` is the target structure type (like `Option` or List)
+
+This `bind` operation looks similar to the other ones you've seen so far, if you put
+them all together `Monad` has the following operations:
+
+```lean,ignore
+class Monad (f : Type u → Type v) extends Applicative f, Bind f where
+  pure {α : Type u} : α → f α
+  map : {α β : Type u} → (α → β) → f α → f β
+  seq : {α β : Type u} → f (α → β) → (Unit → f α) → f β
+  bind : {α β : Type u} → f α → (α → f β) → f β
+  ...
+```
+
+Notice `Monad` also contains `pure` it must also have a "default" way to wrap a value in the
+structure.
+
+The `bind` operator is similar to the applicative `seq` operator in that it chains two operations,
+with one of them being function related.
+
+Notice that `bind`, `seq` and `map` all take a function of some kind, and then a structure wrapped
+over a type `α`, namely `f α`, and then produce a structure wrapping a type `β`, namely `f β`. They
+just vary in what the function looks like. For the functor, the function is a normal pure function
+because it has the type `(α → β)`. For applicatives, the function is still pure, but wrapped in the
+structure `f (α → β)`. Now with monads, the function argument takes a "pure" input `α` but produces
+an output in the structure, `(α → f β)`
+
+## Basic Monad Example
+
+Just as `Option` is a functor and an applicative functor, it is also a monad!
+Let's start with how `Option` implements the Monad type class.
+
+-/
+instance : Monad Option where
+  pure := Option.some
+  bind := Option.bind
+/-!
+
+where:
+
+```lean,ignore
+def Option.bind : Option α → (α → Option β) → Option β
+  | none,   _ => none
+  | some a, f => f a
+```
+
+> **Side note**: this function definition is using a special shorthand syntax in Lean where the `:=
+match a, b with` code can be collapsed away. To make this more clear consider the following simpler
+example, where `Option.bind` is using the second form like `bar`:
+
+-/
+def foo (x : Option Nat) (y: Nat) : Option Nat :=
```

Comment:
Please add space between `y` and `:`. There are many other instances of this issue.

Reply:
Fixed: https://github.com/leanprover/lean4/pull/1505/commits/ed117c1d35e45fadd510a7c2223d3e93b46749c7